### PR TITLE
[Distributed Strategy] Add Refined Recompute support for DeepEP combine overlap

### PR DESCRIPTION
**Pull request by wangbingguang2026**
### PR Category
Distributed Strategy

### PR Types
New features

### Description
在 DeepEP combine 异步操作中引入 Refined Recompute（RR）机制，优化 MoE 模型训练中 recompute 阶段的通信开销。

#### 背景
当前 DeepEP combine 与 shared expert 的 overlap 计算在 recompute 阶段会重复执行通信操作，造成不必要的性能损耗。通过 Refined Recompute 机制，在第一次 forward 时缓存通信结果，在 recompute 的第二次 forward 中直接复用缓存结果，避免重复通信。

#### 主要改动
- **fused_a2a.py**: 新增 `DeepEPCombineAsyncFunctor`（PyLayer）和 `DeepEPCombineAsyncRefinedRecompute` 类，实现两次 forward 的分离逻辑；修改 `fused_combine` 函数支持 RR 分支
- **token_dispatcher.py**: 在 `_DeepEPManager.combine` 方法中集成 RR 逻辑，按需延迟初始化 `DeepEPCombineAsyncRefinedRecompute` 实例
- **moe_layer.py**: 添加 `rr_recompute_update` 方法，根据 `recompute_modules` 配置中的 `moe_combine` 项动态控制 RR 开关
- **transformer_layer.py**: 在层初始化阶段调用 `rr_recompute_update`，传入 recompute 上下文信息

#### 使用方式
在 TransformerConfig 的 `recompute_modules` 中添加 `"moe_combine"` 即可启用。支持 list 模式（全部层启用）和 dict 模式（指定前 N 层使用 recompute，其余层使用 RR）。

### 是否引起精度变化
否。RR 机制不改变计算结果，仅优化 recompute 阶段的执行策略。已分别验证在 Full Recompute 和 Selective Recompute 场景下使用 RR DeepEP Combine，前10个 step 的 loss 和 global_norm 可以逐位对齐。

### CODE REVIEW

File: src/paddlefleet/transformer/moe/fused_a2a.py [hunk 1]
```diff
@@ -517,6 +520,107 @@ def backward(ctx, grad_output, *fn_out_grads):
...
+
+    def __init__(self):
+        """__init__"""
+        self._hold_tensors_queue = queue.Queue()
+        global_rr_queue_log.update(self._hold_tensors_queue, "DeepEPCombineAsync")
+
+    def forward(self, x, group, states, *fn_args, fn):
+        """forward"""
+        tracer = framework._dygraph_tracer()
+        is_first_fwd = not tracer._has_grad
```
[the remark this answers is not in the document]
Comment (wangbingguang2026) [reply src/paddlefleet/transformer/moe/fused_a2a.py:577]: 两者等价，不需要修改。DeepEPCombineAsync 的调用处也是 is_first_fwd=not framework._dygraph_tracer()._has_grad。

**Review by LiYuRio — DISMISSED** [review]

File: src/paddlefleet/transformer/moe/fused_a2a.py [hunk 2]
```diff
@@ -517,6 +520,109 @@ def backward(ctx, grad_output, *fn_out_grads):
...
+        """Forward pass of fused combine with overlap, get cached output directly."""
+        combined_x = hold_tensors["res_output"]
+
+        # Re-run fn with grad tracking to build backward graph and obtain bwf
+        ctx.bwf, fn_out = manual_backward(fn, False, *fn_args)
+
+        ctx.handle = states["handle"]
+        ctx.group = group
+
+        return (combined_x,) + fn_out  # noqa: RUF005
```
[the remark this answers is not in the document]
Comment (wangbingguang2026) [reply src/paddlefleet/transformer/moe/fused_a2a.py:546]: second_fwd 中 `DeepEPCombineAsyncFunctor.forward` 不需要再调用 `wait_for_deepep`，因为通信已在 `_first_fwd` 的 line 611 完成等待。`states["handle"]` 是同一个对象，wait 完成后 handle 仍保留路由元数据（哪些 token 发往哪个 expert），这与原始 `DeepEPCombineAsync` 的模式完全一致——它也是在 forward 中 wait 后、在 backward 中继续使用 `ctx.handle` 发起反向通信。

代码 line 609-610 已有注释说明此设计意图。

File: src/paddlefleet/transformer/moe/fused_a2a.py [hunk 3]
```diff
@@ -517,6 +520,109 @@ def backward(ctx, grad_output, *fn_out_grads):
...
+        )
+
+    def forward(self, x, group, states, *fn_args, fn):
+        """forward"""
+        tracer = framework._dygraph_tracer()
+        is_first_fwd = not tracer._has_grad
+        if is_first_fwd:
+            fwd_output, fn_out = self._first_fwd(x, group, states, fn, *fn_args)
+            self._hold_tensors_queue.put({"res_output": fwd_output.detach()})
+            return (fwd_output, *fn_out)
```
[the remark this answers is not in the document]
Comment (wangbingguang2026) [reply src/paddlefleet/transformer/moe/fused_a2a.py:584]: 已在最新提交中添加注释说明：`_first_fwd` 在 `@paddle.no_grad()` 下运行，返回的 tensor 无梯度。反向计算图在第二次 forward（recompute pass）中通过 `DeepEPCombineAsyncFunctor` 重建，调用方不应依赖 first_fwd 返回值的梯度。

这是 Refined Recompute 的标准模式：first_fwd 只做通信+缓存，second_fwd 重建计算图以支持 backward。

File: src/paddlefleet/transformer/moe/moe_layer.py [hunk 4]
```diff
@@ -432,6 +433,57 @@ def __init__(
...
+            (not in_full_recompute)
+            and (not in_mlp_recompute)
+            and self.use_rr_deepep_combine
+        ):
+            raise ValueError(
+                "Enabling rr for moe_combine is meaningless when neither full_recompute "
+                "nor mlp_recompute is active."
+            )
+
     def _init_expert_parallel(self):
```
[the remark this answers is not in the document]
Comment (wangbingguang2026) [reply src/paddlefleet/transformer/moe/moe_layer.py:485]: `TransformerLayerWithOverlap` 使用 overlap scheduler 单独管理 shared expert 计算，不走 `combine_overlap_handle` 路径，因此 `compute_combine` 不传 `use_rr_deepep_combine` 是 by design（见 `moe_layer.py:908-910` 的注释说明）。

在该场景下，即使 `use_rr_deepep_combine=True` 被设置，它也不会被 `compute_combine` 消费，不会产生运行时错误或副作用。RR 机制针对的是 `fusion_moe_forward` 中通过 `combine_overlap_handle` 做 combine+shared_expert overlap 的常规路径（非 overlap scheduler 路径）。

`TransformerLayerWithOverlap` 的 full_recompute 是由 overlap scheduler 自身的调度机制管理的，不需要 RR 介入 combine 阶段。

**Comment by wangbingguang2026** [comment]
## 回复最新 Code Review（2026-05-21 09:35:57）

### 🟡 `moe_ep_barrier` 参数未透传

确认这是一个存量问题，**并非本 PR 引入**。原始 `DeepEPCombineAsync.forward`（line 490-494）同样未透传 `moe_ep_barrier`，调用 `fused_combine_forward_func` 时使用默认值 `True`。RR 路径与非 RR 的 async 路径行为完全一致。

这是因为 `DeepEPCombineAsync`/RR 路径仅在 `combine_overlap_handle is not None`（即 shared expert overlap 场景）下触发，该场景下 barrier 语义与普通 combine 不同，当前硬编码 `True` 是安全的。如需对齐，可在后续 PR 中统一修复两条路径。

### ❓ dict mode 下 `{"moe_combine": 0}` 全部启用 RR

这是**预期行为**，与 PaddleFleet 中 `need_recompute_in_first_n` 的标准语义一致：

- `recompute_num_layers=N` 表示"前 N 层使用 full recompute"
- 当 N=0 时，没有层使用 full recompute → 所有层启用 RR
- 等价于 list 模式（`recompute_modules=["moe_combine"]`）的全局启用效果

实际使用中，用户使用 dict 模式是为了差异化控制（如 `{"moe_combine": 10}` 表示前 10 层 recompute、其余层用 RR）。`{"moe_combine": 0}` 虽然合法，但在语义上等同于 list 模式，用户一般不会这样配置。此外，外层 guard（line 477-485）已确保 recompute_granularity 必须开启，不存在无 recompute 场景下误启用 RR 的风险。

**Review by LiYuRio — DISMISSED** [review]

**Review by LiYuRio — DISMISSED** [review]

File: src/paddlefleet/transformer/moe/token_dispatcher.py [hunk 5]
```diff
@@ -595,15 +598,27 @@ def combine(
         hidden_states: paddle.Tensor,
```
[the remark this answers is not in the document]
Comment (wangbingguang2026) [reply src/paddlefleet/transformer/moe/token_dispatcher.py:598]: `_rr_fusedcombined` 的生命周期与所属 `_DeepEPManager` 实例（即模型）一致，模型存在则实例存在，模型释放则自然回收，不存在泄漏。

当前 `use_rr_deepep_combine` 在 init 阶段由配置确定，运行期间不会动态切换，这是有意的设计约束。`_hold_tensors_queue` 在每个 step 内严格遵循 put/get 配对（first_fwd put → second_fwd get），`global_rr_queue_log.check()` 会在 step 结束时校验队列为空。

如果未来需要支持动态开关，届时再引入清理逻辑即可；当前无需提前增加复杂度。

**Comment by wangbingguang2026** [comment]
关于补充多卡集成测试的建议：

认同这个方向。当前 PR 已通过多卡环境手动验证（EP=8 场景下 loss 对齐、通信次数减半确认），核心逻辑也有单元测试覆盖。

完整的 `MoELayer` → `TransformerLayer` 链路多卡回归测试涉及 EP 并行资源调度，计划作为 follow-up 在后续 PR 中补充，不阻塞当前功能合入。

**Review by LiYuRio — DISMISSED** [review]

**Comment by wangbingguang2026** [comment]
re-run all-failed

**Review by LiYuRio — APPROVED** [review]
